### PR TITLE
Processing data in models

### PR DESCRIPTION
**Pull request by clin2on3mun**
in this milestone :
Create models
- Create model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png).
- Set up associations between models.
  - Remember that author_id column in posts table should be the foreign_key for the users table.
- Create custom methods
    - Users
       - A method that returns the 3 most recent posts for a given user.
    - Posts
      - A method that updates the posts counter for a user.
      - A method which returns the 5 most recent comments for a given post.
    - Comments
       - A method that updates the comments counter for a post.
    - Likes
       - A method that updates the likes counter for a post.

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_many :likes
+  has_many :comments
+
+  after_save :update_post_comment
+  def update_post_counter
+    author.increment!(:posts_counter)
+  end
```
Comment (algerina) [inline app/models/post.rb]: - [ ] Well done using `after_save ` to call the `#update_post_counter ` method every time a new post has been saved to database, I  just recommend to make the update_post_counter method private.


`after_save :update_post_comment

private

  def update_post_counter
    author.increment!(:posts_counter)
  end`

File: app/models/like.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class Like < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
+
+  after_save :update_post_likes_counter
+  def update_post_likes_counter
+    posts.increment!(:likes_counter)
+  end
```
Comment (algerina) [inline app/models/like.rb]: - [ ] Well done using after_save  to call the #update_post_likes_counter  method every time a new post has been saved to database, I just recommend to make the update_post_likes_counter method private.

File: app/models/comment.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  belongs_to :post
+
+  after_save :update_post_comment
+  def update_post_comment
+    posts.increment!(:comments_counter)
+  end
```
Comment (algerina) [inline app/models/comment.rb]: - [ ] Well done using after_save  to call the #update_post_comment  method every time a new post has been saved to database, I just recommend to make the update_post_comment method private.

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: 'author_id'
+  has_many :likes
```
Comment (algerina) [inline app/models/user.rb]: - [ ] Well done setting up the associations between models, I recommend to add the foreign key for likes.

`has_many :likes, foreign_key: 'author_id'`

**Review by algerina — CHANGES_REQUESTED** [review]
Hi @clin2on3mun  ,

Great job so far!  You just have a small issue to work on.

May I highlight the following: 
-  Well done setting up associations between model according to the ERD diagram✔
- The code is clean and well organized✔️



## CHANGES REQUIRED:muscle::clap:


Check the comments under the review.



 
 ## Optional suggestions

 - Please check the comment under the review.

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._
Cheers and Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread and tag me @algerina if something is not 100% clear.

#### _**Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.**_
------------------------------------------------------------------------------------------------------------------------------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using this [form.](https://airtable.com/shrQAqnBwek5a0O0s)
___
_If anything seems to be unclear to you, reach out and ask anything by commenting on this Pull Request and mention me as [Algerina](https://github.com/algerina), I will receive a notification and will respond._

[GitHub](https://github.com/algerina) | [LinkedIn](https://linkedin.com/in/amel-khiri) | [Twitter](https://twitter.com/AmelQahwadji)

_Reviewed by [Amel Khiri ](https://github.com/algerina)_

**Comment by prolajumokeoni** [comment]
Hi @clin2on3mun, please fix the rubocop linter errors

**Comment by clin2on3mun** [comment]
> Hi @clin2on3mun, please fix the rubocop linter errors



> Hi @clin2on3mun, please fix the rubocop linter errors

@prolajumokeoni sorry, i have fixed the linters. thank you!

**Review by prolajumokeoni — APPROVED** [review]
Hi @clin2on3mun 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
## Highlights
- Associations set between models correctly ✅ 
  

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**